src/main/java/com/gooddata/executeafm/result/Data.java

```diff
@@ -0,0 +1,108 @@
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.stream.Collectors;
+
+import static com.fasterxml.jackson.databind.JsonMappingException.from;
+import static java.util.Spliterators.spliteratorUnknownSize;
+import static java.util.stream.StreamSupport.stream;
+
+/**
+ * Data of {@link ExecutionResult}, can be of three basic kinds - {@link #NULL}, list and simple value.
+ */
+@JsonDeserialize(using = Data.DataDeserializer.class)
+public interface Data {
+
+    Data NULL = new Data() {
+        @Override
+        public boolean isList() {
+            return false;
+        }
+
+        @Override
+        public boolean isValue() {
+            return false;
+        }
+
+        @Override
+        @JsonValue
+        public String textValue() {
+            return null;
+        }
+    };
+
+    /**
+     * @return true if this instance is of kind list, false otherwise
+     */
+    boolean isList();
+
+    /**
+     * @return true if this instance is of kind value, false otherwise
+     */
+    boolean isValue();
+
+    /**
+     * @return true if this instance is the same as {@link #NULL}, false otherwise
+     */
+    default boolean isNull() {
+        return this == NULL;
+    }
+
+    /**
+     * @return text data value, throws exception for data which can't be represented as text
+     */
+    String textValue();
+
+    /**
+     * @return this instance cast to List, may throw exception if tthis instance is not of kind list.
+     */
+    default List<Data> asList() {
+        throw new UnsupportedOperationException("This is not a list");
+    }
+
+    class DataDeserializer extends JsonDeserializer<Data> {
+        @Override
+        public Data deserialize(final JsonParser jp, final DeserializationContext ctxt) throws IOException {
+            final JsonNode root = jp.readValueAsTree();
+            if (root.isArray()) {
+                final List<Data> list = stream(spliteratorUnknownSize(root.elements(), Spliterator.ORDERED), false)
+                        .map(elem -> {
+                            try {
+                                final Class<? extends Data> clazz = elem.isArray() ? DataList.class : DataValue.class;
+                                return ctxt.readValue(elem.traverse(jp.getCodec()), clazz);
+                            } catch (IOException e) {
+                                throw new UncheckedIOException(e);
+                            }
+                        }).collect(Collectors.toList());
+                return new DataList(list);
+            } else if (root.isTextual()) {
+                return new DataValue(root.textValue());
+            } else if (root.isNull()) {
+                return NULL;
+            } else {
+                throw from(jp, "Unknown value of type: " + root.getNodeType());
```

Comment:
I would add `JsonMappingException.` for better readability.

Reply:
added